tao_compiler/mlir/disc/IR/hlo_disc_ops.cc

```diff
@@ -269,6 +270,118 @@ LogicalResult SparseReshapeOp::verify() {
+  // TODO(lanbo.llb): Handle corner case when dense_rows == 0
+  Value dense_rows = builder.create<arith::IndexCastOp>(
+      loc, builder.getIndexType(),
+      builder.create<tensor::ExtractOp>(loc, operands[2], idx_zero));
+  // outputs:
+  // 0: output_indices
+  // 1: output_values
+  // 2: empty_row_indicator
+  // 3: reverse_index_map
+  // 4: output_elements(scalar)
+
+  // However N_full can only be determined by the values from indices
+  // so we allocate a output_indices/output_values to max possible size, which
+  // is `N + dense_rows`
+  Value N_full = builder.create<arith::AddIOp>(loc, N, dense_rows);
+
+  // output indices: {N + dense_rows, rank}
+  SmallVector<Value, 2> output_indices_shape_values;
+  output_indices_shape_values.push_back(N_full);
+  output_indices_shape_values.push_back(rank);
+  Value output_indices_shape =
+      builder.create<tensor::FromElementsOp>(loc, output_indices_shape_values);
+  reifiedReturnShapes.push_back(output_indices_shape);
+
+  // output values: {N + dense_rows}
+  SmallVector<Value, 2> output_values_shape_values;
+  output_values_shape_values.push_back(N_full);
+  Value output_values_shape =
+      builder.create<tensor::FromElementsOp>(loc, output_values_shape_values);
+  reifiedReturnShapes.push_back(output_values_shape);
+
+  // empty indicator: {dense_rows}
+  SmallVector<Value, 1> empty_row_indicator_shape_values;
+  empty_row_indicator_shape_values.push_back(dense_rows);
+  Value empty_row_indicator_shape = builder.create<tensor::FromElementsOp>(
+      loc, empty_row_indicator_shape_values);
+  reifiedReturnShapes.push_back(empty_row_indicator_shape);
+
+  // reverse index map: {N}
+  SmallVector<Value, 1> reverse_index_map_shape_values;
+  reverse_index_map_shape_values.push_back(N);
+  Value reverse_index_map_shape = builder.create<tensor::FromElementsOp>(
+      loc, reverse_index_map_shape_values);
+  reifiedReturnShapes.push_back(reverse_index_map_shape);
+
+  // output_elements
+  Value idx_one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  SmallVector<Value, 1> output_elements_shape_values;
+  output_elements_shape_values.push_back(idx_one);
+  Value output_elements_shape =
+      builder.create<tensor::FromElementsOp>(loc, output_elements_shape_values);
+  reifiedReturnShapes.push_back(output_elements_shape);
+
+  return success();
+}
+
+LogicalResult SparseFillEmptyRowsOp::verify() {
+  // index 0
+  auto indices_type = this->indices().getType().dyn_cast<ShapedType>();
+  // index 1
+  auto values_type = this->values().getType().dyn_cast<ShapedType>();
+  // index 2
+  auto dense_shape_type = this->dense_shape().getType().dyn_cast<ShapedType>();
+  // index 3
+  auto default_value_type =
+      this->default_value().getType().dyn_cast<ShapedType>();
+
+  if (!indices_type || !values_type || !dense_shape_type ||
+      !default_value_type) {
+    return failure();
+  }
+
+  if (dense_shape_type.getRank() != 1) {
```

Comment:
do we also need the check `dense_shape_type ` is actually a static shape type since we do not support dynamic rank?

Reply:
done